tests/matchers.py

```diff
@@ -44,6 +44,39 @@ def __repr__(self):
         return repr_(self, self.repr_attrs)
 
 
+class InstanceOf(Matcher):
+    """Matches any instance of the given class with the given attrs.
+
+    As with Python's builtin isinstance() `class_` can be either a single class
+    or a tuple of classes (in which case the matcher will match instances of
+    *any* of the classes in the tuple).
+
+    If no kwargs are given then the matcher will match instances of the given
+    class(es) without checking any of the instance's attributes.
+    """
+
+    def __init__(self, class_, **kwargs):
+        self.class_ = class_
+        self.with_attrs = kwargs
+
+    @property
+    def repr_attrs(self):
+        if self.with_attrs:
+            return ("class_", "with_attrs")
+
+        return ("class_",)
```

Comment:
In most matchers `repr_attrs` is not dynamic so it's just a class attribute:

```python
class MyMatcher:
    repr_attrs = ("foo", "bar")
    ...
```

In cases like this where we want `repr_attrs` to vary dynamically, an `@property` works well.